internal/blocksync/reactor.go

```diff
@@ -489,94 +485,112 @@ func (r *Reactor) poolRoutine(ctx context.Context, stateSynced bool, blockSyncCh
 			//
 			// TODO: Uncouple from request routine.
 
-			// see if there are any blocks to sync
-			first, second := r.pool.PeekTwoBlocks()
-			if first == nil || second == nil {
-				// we need both to sync the first block
+			newBlock, verifyBlock := r.pool.PeekTwoBlocks()
+
+			if newBlock == nil || verifyBlock == nil {
 				continue
 			} else {
-				// try again quickly next loop
 				didProcessCh <- struct{}{}
 			}
 
-			firstParts, err := first.MakePartSet(types.BlockPartSizeBytes)
-			if err != nil {
+			newBlockParts, err2 := newBlock.MakePartSet(types.BlockPartSizeBytes)
+			if err2 != nil {
 				r.logger.Error("failed to make ",
-					"height", first.Height,
-					"err", err.Error())
+					"height", newBlock.Height,
+					"err", err2.Error())
 				return
 			}
 
 			var (
-				firstPartSetHeader = firstParts.Header()
-				firstID            = types.BlockID{Hash: first.Hash(), PartSetHeader: firstPartSetHeader}
+				newBlockPartSetHeader = newBlockParts.Header()
+				newBlockID            = types.BlockID{Hash: newBlock.Hash(), PartSetHeader: newBlockPartSetHeader}
 			)
 
-			// Finally, verify the first block using the second's commit.
-			//
-			// NOTE: We can probably make this more efficient, but note that calling
-			// first.Hash() doesn't verify the tx contents, so MakePartSet() is
-			// currently necessary.
-			if err = state.Validators.VerifyCommitLight(chainID, firstID, first.Height, second.LastCommit); err != nil {
-				err = fmt.Errorf("invalid last commit: %w", err)
-				r.logger.Error(
-					err.Error(),
-					"last_commit", second.LastCommit,
-					"block_id", firstID,
-					"height", first.Height,
-				)
+			if r.lastTrustedBlock != nil {
+				err := VerifyNextBlock(newBlock, newBlockID, verifyBlock, r.lastTrustedBlock.block, r.lastTrustedBlock.commit, state.NextValidators)
```

Comment:
Actually I am not 100% sure that the way it was done before was correct. 
In the previous scenario, `state` is the state existing at height `H `and then we want to verify the `first` block at height` H + 1`. So one should verify the `first` block with the validator set for the height of that block. But the `validator` field of the state relates to the validators that validate that particular state, not necessarily the set that will be used to validate the state at the next height.

So here I take the validator set of` H + 1` to verify `newBlock` , and at the same time [here](https://github.com/tendermint/tendermint/blob/4b83be646c140fb50391acc40384dcb1cd2e10e6/internal/blocksync/verify.go#L95): I use it to verify that these validators indeed signed the commit of `verifyBlock.lastCommit`(on height H + 2). It is assumed that we trust the state but we need the last commit of verifyBlock to verify newBlock. This was the only way I could think of to at least somehow increase the trust in verifyBlock.

Reply:
OK, so if you followed your explanation correctly, the old code, instead of `state.Validators.VerifyCommitLight(...)` should have been doing `state.NextValidators.VerifyCommitLight(...)` all along. Is that so?

Reply:
After `ApplyBlock`, `state.LastBlockHeight` is `NewBlock.Height` and state is thus for `NewBlock.Height + 1`

Reply:
I confirmed this in the doc, `ApplyBlock` returns state for H + 1 and changed this  to use` state.Validators`.